### ISSUE TITLE
[Discussion] : Usage of Open Source LLM ?

### ISSUE THREAD

**Issue by kshitijrajsharma** [rowheat02/osm-gpt#1]
I Liked your project , One  similar project https://trident.yuiseki.net/ , https://github.com/yuiseki/TRIDENT which follows Gpt3 and overpass API and its pretty great actually  

As we know charges for API usage will be massive , I was thinking if we can leverage power of open source LLM  and can train them , Do you have any ideas to share on this topic  ? I will be happy to contribute !
Since project still in development if we can use open source llm with training data we can create a good base 
 Wanted to hear thoughts on this  
 cc : @rowheat02

**Comment by konishon** [comment]
I really like this idea

**Comment by rowheat02** [comment]
@kshitijrajsharma  Thank you for showing interest and offering to contribute! I really appreciate your enthusiasm. Your idea of using open-source LLMs to save on API costs is excellent. We can explore options like Meta's Llama2, a recently released LLM, or other advanced models.

I'd love to hear your ideas and suggestions on how we can acquire a sufficient amount of OSM query data to train the model effectively. Please share any potential sources or data collection methods, you think would be beneficial. 
Let's collaborate and make this project cost-effective!

**Comment by kshitijrajsharma** [comment]
We can get sufficient query , that won't be a problem at the end its a pair between language keywords with overpass query ! 
For the model itself 
My Initial Thoughts :  
**llama : Released By Meta**
https://ai.meta.com/llama/ 
I haven't tested this out yet intensively but looks like a promising model as they advertise , They say its free for researchers and commercial use but I am concerned with their license , Lot of people say its not open source : https://blog.opensource.org/metas-llama-2-license-is-not-open-source/ 
I have tried their LLaMA2-70B Model , Not good as chatgpt , and produces random query . I wonder if it can be retrained then this could work 
![image](https://github.com/rowheat02/osm-gpt/assets/36752999/affec23a-8c68-4ac6-bd96-18bf2d4bd104)

Another option is stable LM 
https://github.com/Stability-AI/StableLM
And this one : 
https://github.com/abetlen/llama-cpp-python

**Comment by kshitijrajsharma** [comment]
Update on LLama model training : 

Here is the sample dataset that can be used to train LLama with RLHF 

https://huggingface.co/datasets/HuggingFaceH4/stack-exchange-preferences 

Now we need similar training dataset to be prepared for overpass questions and query !

**Comment by rowheat02** [comment]
Thank you @kshitijrajsharma for sharing the sample dataset and references for training LLama with RLHF, as well as the insight into the challenges ahead. The stack-exchange-preferences dataset seems valuable, and it's promising that LLama can potentially run on an 8GB GPU. Although generating a similar training dataset for overpass questions will require effort, we are now equipped with the knowledge to proceed. I appreciate the progress made so far and believe we are ready to collect data for testing fine-tuning. Collaboration within the community will be vital in overcoming the challenge and achieving success. Let's continue to build on these findings and unlock LLama's full potential!

**Comment by kshitijrajsharma** [comment]
Yes ! There is a option for collecting data from the tool within though ! Need to check licensing for it ! Now I am looking at training datasets , let me know if you find any options

**Comment by orkutmuratyilmaz** [comment]
Hello and thanks for this beautiful repo:)

If you'd like to integrate a self hosted solution, please consider [Ollama](https://github.com/jmorganca/ollama/issues).

Best,
Orkut

**Comment by simonprovost** [comment]
The Llama 4 herd is out! Maybe it would be a nice try!
